feat(blog): langchain4j (#2635)
kestra-io/docs: content/blogs/rag-with-gemini-and-langchain4j.md

---
title: "Retrieval Augmented Generation (RAG) with Google Gemini AI and Langchain4J"
description: Create a Retrieval Augmented Generation pipeline with Google Gemini AI and the Langchain4J plugin.
date: 2025-06-10T13:00:00
category: Solutions
author:
  name: Loïc Mathieu
  image: lmathieu
  role: Lead Developer
image: /blogs/TODO.png
---

Generative AI tools are great. However, relying purely on generative models can lead to outputs that feel generic, inaccurate, or outdated. This is where **Retrieval-Augmented Generation (RAG)** comes in, combining the creativity of Generative AI with real-time, accurate context sourced from custom data.

> We just introduced the [Langchain4J plugin](https://blog-lanchain4j.kestra-io.pages.dev/plugins/plugin-langchain4j) that allows creating complex Generative AI workflows in an AI provider-agnostic way. This plugin also provides all the tasks needed to create a RAG pipeline. 

Reviewer: ```suggestion
We just introduced the [Langchain4J plugin](https://blog-lanchain4j.kestra-io.pages.dev/plugins/plugin-langchain4j) that allows users to create complex Generative AI workflows in an AI provider-agnostic way. This plugin also provides all the tasks needed to create a RAG pipeline. 
```

In this post, you'll learn how to build an advanced RAG pipeline using Google Gemini AI and Kestra’s new Langchain4J plugin, which leverages precise, context-aware AI generation. We'll cover document ingestion, embedding creation, retrieval strategies, and finally, demonstrate a full end-to-end RAG example.

## Creating Powerful AI Workflows with Kestra’s Langchain4J

The new [Langchain4J plugin](https://blog-lanchain4j.kestra-io.pages.dev/plugins/plugin-langchain4j) simplifies the implementation of RAG, handling document ingestion, embeddings creation, retrieval, and generation tasks.

Here's a high-level overview of the workflows you'll create:

- **Document Ingestion and Embedding Creation:** You’ll ingest documents, splitting them into manageable segments and generating embeddings.
- **Embedding Storage:**  Store embeddings in a vector store, optimizing retrieval.
- **Retrieval and Augmented Generation:** Build workflows that retrieve the most relevant embeddings based on user prompts and generate context-rich, accurate responses with Gemini AI.

> ## Embedding creation via document ingestion

Reviewer: ```suggestion
## Create Embeddings through Document Ingestion
```

    embeddings: #2
      type: io.kestra.plugin.langchain4j.embeddings.KestraKVStore
    fromExternalURLs: #3
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-22.md
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-21.md
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-20.md
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-19.md
    drop: true #4
```

1. We use the `io.kestra.plugin.langchain4j.provider.GoogleGemini` model provider to use the embedding model from **Google Gemini**. We must use the same model at generation time. Kestra supports a large variety of large language models (LLM) and more will be supported soon.
2. We use the `io.kestra.plugin.langchain4j.embeddings.KestraKVStore` embedding store. This is a convenient store that will store embeddings inside a [KeyValue store](https://kestra.io/doc/concepts/kv-store) and load them all in memory at generation time. For a large number of documents, you would typically use a vector database instead like PGVector or Elasticsearch.

> 3. We use `fromExternalURLs` to defines a list of documents to ingest from external URLs, here the blog posts for Kestra release 0.19 to 0.22. We will go into details for other ways to define documents to ingest later.

Reviewer: ```suggestion
3. We use `fromExternalURLs` to define a list of documents to ingest from external URLs; here, the blog posts for Kestra releases 0.19 to 0.22. We will go into detail for other ways to define documents to ingest later.
```

4. We use `drop: true` so that the embedding store is recreated each time the flow is executed.

After executing the flow, you will be able to see a new KV store entry with the serialized form of the computed embeddings.

![Embedding Store KV Entry](/blogs/rag-with-gemini-and-langchain4j/embedding-store-kv.png)

### Defines document from multiple sources

You can define documents from multiple sources:
- `fromPath`: from a working directory path, usually used withing a [WorkingDirectory](https://kestra.io/plugins/core/flow/io.kestra.plugin.core.flow.workingdirectory) task, each file in the directory will create a document.
- `fromInternalURIs`: from a list of internal storage URIs.
- `fromExternalURLs`: from a list of external URLs.
- `fromDocuments`: from a list of documents defined inside the task itself.

Document metadata allows defining additional information to the large language model, some metadata is automatically added to documents at ingestion time. For example, the URL of the document if using `fromExternalURLs` or the name of the file if using `fromPath`.
You can set additional metadata via the `metadata` property.

### Advanced document splitting

> Documents are split in segments to feed the large language model. Document splitting is an important step as each segment will create an embedding vector, so the embedding retrieval performance and accuracy will depend on how documents are split.

Reviewer: ```suggestion
Documents are split into segments to feed the large language model. Document splitting is an important step, as each segment will create an embedding vector, so the embedding retrieval performance and accuracy will depend on how you split the documents.
```

```yaml
id: rag-completion
namespace: company.team

inputs:
  - id: prompt
    type: STRING
    defaults: What's new in Kestra 0.22?

tasks:
  - id: completion
    type: io.kestra.plugin.langchain4j.rag.Chat
    embeddings: #1
      type: io.kestra.plugin.langchain4j.embeddings.KestraKVStore
      kvName: ingest-documents-embedding-store
    chatProvider: #2
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-2.5-flash-preview-05-20
      apiKey: "{{secret('GEMINI_API_KEY')}}"
    embeddingProvider: #3
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-embedding-exp-03-07
      apiKey: "{{secret('GEMINI_API_KEY')}}"
    contentRetrieverConfiguration: #4
      maxResults: 3
      minScore: 0.5
    prompt: "{{ inputs.prompt }}" #5
```

1. We're using here the same embedding store as the one used in the `ingest-documents` flow. We must set the name of the KV entry as we are not in the same flow and the name of the KV entry is by default the name of the flow suffixed by `embedding-store`.

> 2. We're using here the `io.kestra.plugin.langchain4j.provider.GoogleGemini` large language model provider configured to use the Google Gemini Flash 2.5 model.

Reviewer: ```suggestion
2. We're using the `io.kestra.plugin.langchain4j.provider.GoogleGemini` large language model provider configured to use the Google Gemini Flash 2.5 model.
```

3. We're using here the `io.kestra.plugin.langchain4j.provider.GoogleGemini` large language model provider for embeddings, this must be the same as the one used to ingest documents into the embedding store.

> 4. We configure the content retriever to return 3 results and filter results with a minimal score of 0.5 to avoid having non-accurate results returned by the embedding store.

Reviewer: ```suggestion
4. We configure the content retriever to return three results and filter them with a minimal score of 0.5 to avoid having inaccurate results returned by the embedding store.
```

Here are some of the most interesting new features:

1.  **Apps: Custom UIs for Your Flows (Kestra 0.20)**
    This feature allows users to build custom interfaces for interacting with Kestra workflows. It democratizes access to workflows by providing simple forms, output displays, and approval buttons, enabling non-technical business users to trigger, pause, or submit data to automated processes without needing to understand the underlying code. Flows act as the backend, while Apps serve as the frontend.
    *   **Source:** [Kestra 0.20 adds SLAs, Invites, User-Facing Apps, Isolated Storage and Secrets per Team, and Transactional Queries](https://kestra.io/blogs/2024-12-03-release-0-20#apps)

[...]
```

### How it works

The prompt is first used to create an embedding vector. This vector will be used to search for the most relevant segments into the embedding store. Here we ask to retrieve three documents with a minimal score of 0.5.

Then, these documents are sent into the context window of the LLM with the prompt for generation.

> At ingestion time, each document will be indexed with metadata, including the URL of the document as they have been retrieved via external URLs. The LLM is then capable of including these URLs as the source of the information used for the generation.

Reviewer: ```suggestion
At ingestion time, each document will be indexed with metadata, including the document URL, as the external URLs property retrieved it. The LLM can then include these URLs as the source of the information used for the generation.
```
